app/server/lib/ActiveDoc.ts

```diff
@@ -987,8 +989,17 @@ export class ActiveDoc extends EventEmitter {
           throw new ApiError("Document is shutting down, archiving aborted", 500);
         }
         const file = await attachmentFileManager.getFile(attachment.fileIdent);
-        const fileHash = attachment.fileIdent.split(".")[0];
-        const name = `${fileHash}_${attachment.fileName}`;
+        const fileIdentParts = attachment.fileIdent.split(".");
+        const fileHash = fileIdentParts[0];
+        const fileIdentExt = path.extname(attachment.fileIdent);
+        const fileNameExt = path.extname(attachment.fileName);
+        const fileNameNoExt = path.basename(attachment.fileName, fileNameExt);
+        // We need to make sure the downloaded attachment's extension matches Grist's internal
+        // file extension, otherwise we can't recreate the file identifier when uploading.
+        const extension = fileIdentExt || fileNameExt;
+        const suffix = extension ? `.${extension}` : "";
+
+        const name = `${fileHash}_${fileNameNoExt}${suffix}`;
```

Comment:
Small tweak to the names of files in downloaded archives (see comment). Shouldn't make any real difference, but covers an edge case.